src/viam/examples/modules/mlmodelservice_tflite.cpp

```diff
@@ -0,0 +1,759 @@
+        // our case, the caller is MLModelServiceServer, which will
+        // copy the data into the reply gRPC proto and then unwind. So
+        // we can avoid copying the data by letting the views alias
+        // the tensorflow tensor buffers and keep the interpreter lock
+        // held until the gRPC work is done. Note that this means the
+        // interpreter lock will remain held until the
+        // inference_result_type object tracked by the shared pointer
+        // we return is destroyed. Callers that want to make use of
+        // the inference results without keeping the interpreter
+        // locked would need to copy the data out of the views and
+        // then release the return value.
+        struct inference_result_type {
+            std::shared_ptr<struct state> state;
+            std::unique_lock<std::mutex> interpreter_lock;
+            named_tensor_views views;
+        };
+        auto inference_result = std::make_shared<inference_result_type>();
+
+        // Walk the outputs per our metadata and emplace an
+        // appropriately typed tensor_view aliasing the interpreter
+        // output tensor buffer into the inference results.
+        for (const auto& output : state->metadata.outputs) {
+            const auto where = state->output_tensor_indices_by_name.find(output.name);
+            if (where == state->output_tensor_indices_by_name.end()) {
+                continue;  // Should be impossible
+            }
+            const auto* const tflite_tensor =
+                TfLiteInterpreterGetOutputTensor(state->interpreter.get(), where->second);
+            inference_result->views.emplace(output.name,
+                                            std::move(make_tensor_view_(output, tflite_tensor)));
+        }
+
+        // The views created in the loop above are only valid until
+        // the interpreter lock is released, so we keep the lock held
+        // by moving the unique_lock into the inference_result
+        // object. The lock itself is only valid as long as the state
+        // is alive, so we move that in too.
+        inference_result->state = std::move(state);
+        inference_result->interpreter_lock = std::move(lock);
+
+        // Finally, construct an aliasing shared_pointer which appears
+        // to the caller as a shared_ptr to views, but in fact manages
+        // the lifetime of the inference_result. When the
+        // inference_result object is destroyed, the lock will be
+        // released and the next caller can invoke the interpreter.
+        auto* const views = &inference_result->views;
+        return {std::move(inference_result), views};
+    }
+
+    struct metadata metadata() final {
+        // Just return a copy of our metadata from leased state.
+        return lease_state_()->metadata;
+    }
+
+   private:
+    struct state;
+
+    std::shared_ptr<state> lease_state_() {
+        // Wait for our state to be valid and then an incrementing
```

Comment:
I forgot a verb there. Changed to:
```
        // Wait for our state to be valid or stopped and then an
        // obtain a shared_ptr to state if valid, incrementing the
        // refcount, or throws if the service is stopped. We don't
        // need to deal with interruption or shutdown because the gRPC
        // layer will drain requests during shutdown, so it shouldn't
        // be possible for callers to get stuck here.
```